### ISSUE TITLE
Decouple write (to disk) from rollup.watch

### ISSUE THREAD

**Issue by jdanyow** [rollup/rollup#3225]
### Feature Use Case

As a dev tooling author I'm using the rollup javascript API to generate bundles in memory. The rollup.rollup API works great for this use case because it separates bundle generation from bundle writing. Would it be possible to make the rollup.watch API match? Maybe if write options are not passed this line could be skipped:

https://github.com/rollup/rollup/blob/8318bb42ff91c510f740af07d8e41c4b3bd99918/src/watch/index.ts#L213

### Feature Proposal

Provide a way to configure rollup.watch such that it does not write the generated bundles to disk.

**Comment by shellscape** [comment]
Thanks for opening an issue. Could you link to your project or any relevant source that would provide context for the request? As this would be a breaking change, maximum information is important.

**Comment by jdanyow** [comment]
Sure, thanks for considering this. I can't link to the repo because it's private but here's the code.

This class is part of a system that bundles chai tests and executes them in headless chrome (puppeteer) or in cloudflare workers. The bundle never needs to be written to disk because it's served from memory with the http module or posted to the cloudflare api to enable testing in the codes true target environment (cloudflare worker).

Specifically, in the `async bundle() {` method below, by the time the watcher's BUNDLE_END event has fired the bundle has been generated and written to disk. I would like to be able to generate myself and avoid the write to disk because the file isn't needed and causes collisions / requires cleanup.

```js
// @ts-check

const EventEmitter = require('events');
const { rollup, watch } = require('rollup');
const glob = require('fast-glob');
const typescript = require('rollup-plugin-typescript2');
const virtual = require('rollup-plugin-virtual');
// const resolve = require('rollup-plugin-node-resolve');
// const commonjs = require('rollup-plugin-commonjs');


/**
 * Bundles specs into a single file using rollup
 */
class Bundler extends EventEmitter {
  /** @type {import('rollup').OutputOptions} */
  outputOptions = { file: 'temp.js', format: 'iife', globals: { 'chai': 'chai' } };

  /**
   * @param {string[]} globs 
   * @param {boolean} watch 
   */
  constructor(globs, watch) {
    super();
    this.globs = globs;
    this.watch = watch;
    this.code = '';
    this.bundled = new Promise(resolve => this.setBundled = resolve);
  }

  async bundle() {
    const inputOptions = await this.getInputOptions();

    if (this.watch) {
      /** @type {import('rollup').RollupWatchOptions} */
      const watchOptions = {
        ...inputOptions,
        output: this.outputOptions
      }
      const watcher = watch([watchOptions]);
      watcher.on('event', async ({ code, result: bundle }) => {
        if (code !== 'BUNDLE_END') {
          return;
        }

        // *** At this point the bundle has been written to disk which is not required for my use case ***

        this.updateCode(bundle);
      });
    } else {
      const bundle = await rollup(inputOptions);
      this.updateCode(bundle);
    }
  }

  /**
   * @param {import('rollup').RollupBuild} bundle
   */
  async updateCode(bundle) {
    const { output } = await bundle.generate(this.outputOptions);
    this.code = output[0].code;
    this.setBundled();
    this.emit('bundled');
  }

  async generateVirtualEntry() {
    /** @type {string[]} */
    const specs = (await Promise.all(this.globs.map(pattern => glob(pattern))))
      // @ts-ignore (flat is not in tslib yet)
      .flat()
      .filter((x, i, a) => a.indexOf(x) === i);
    return specs.map(spec => `import './${spec}';`).join('\n');
  }

  async getInputOptions() {
    const entrySource = await this.generateVirtualEntry();

    /** @type {import('rollup').RollupOptions} */
    const inputOptions = {
      input: 'entry.ts',
      plugins: [
        virtual({ 'entry.ts': entrySource }),
        // @ts-ignore
        typescript({})
      ],
      external: ['chai'],
    };

    return inputOptions;
  }
}

module.exports.Bundler = Bundler;
```

**Comment by PepsRyuu** [comment]
Hey @jdanyow, perhaps this is a use case for Nollup? I had a similar use case and it was one of my driving motivations for developing the project. It provides development bundles for Rollup projects with almost no changes to the config and does not write to disk. There's also an Express middleware that can be added:

https://github.com/PepsRyuu/nollup/blob/master/API.md#dev-middleware

On a side note, here's an example of how I achieved a similar use case: https://github.com/PepsRyuu/preact-skeleton/blob/master/package.json#L6

**Comment by shellscape** [comment]
@PepsRyuu while we're fans of what you're doing with Nollup, it's a tad improper to redirect (advertise?) folks to your project within another project's issues. The OP has several methods of contact on their profile, and that would be a more apt method of communication to tell them about the benefits of Nollup.

I'd like to get this issue back on topic and see if we can't improve Rollup with the OP's help.

**Comment by shellscape** [comment]
@jdanyow I think there's merit to getting rollup to a place where we can support that. _(ping @lukastaegert)_

**Comment by PepsRyuu** [comment]
@shellscape My apologies. I wasn't aware that the rules have changed, and that we weren't allowed to link to third-party plugins and tools, even if they're listed on Rollup sites. Feels a bit weird to pretend that workarounds don't exist. Figured it would be wiser to unblock and solve user problems immediately while a proper solution is being developed in parallel which can take a significant amount of time.

I have removed the post. Will refrain from sharing third-party links in the future.

**Comment by shellscape** [comment]
There's no need to apologize, and there are no official rules about that or third party links. It's just timing. If we'd already evaluated the request and determined that it wasn't a good fit for Rollup or that it was a `wontfix` situation then yes, absolutely a workaround is appropriate.

The issue is pretty fresh, we'd like to work on improving Rollup rather than sending them elsewhere (which more often than not will derail the discussion). That's why I used the phrase "a _tad_ improper," meaning a very small amount. We've sent users to Nollup many many times, but only after working through an issue first.
